### PR TITLE
Merge DBG_printf_gcc and DBG_printf_c99

### PR DESCRIPTION
**Pull request by marler8997**

### CODE REVIEW

**Comment by marler8997** [comment]
Oh...I think I flipped the logic here, I was supposed to remove the `c99` variant...not the `gcc` one.

**Comment by janvorli** [comment]
> Oh...I think I flipped the logic here
I meant it the way you had it before. I think the check for the `__STDC_VERSION__` is not correct and should be removed too. We use variadic macros in corefx using the standard syntax and they work.

**Comment by marler8997** [comment]
So the way it was before, it didn't build on my machine, does that mean we still need to support both?

**Comment by janvorli** [comment]
> So the way it was before, it didn't build on my machine

It didn't build because of the incorrect check for the `__STDC_VERSION__`, I believe. As I've mentioned in my previous comment, if we remove the #ifdef that checks that version, I believe it would build.

**Comment by marler8997** [comment]
Ok I've done that but I had to make 3 more modifications:

1. Use `LPCSTR` instead of `LPSTR` for the filename and format arguments
2. Copy the `__ASSERT_ENTER` macro from the `__GNUC__` ifdef block into where the `__STDC_VERSION__` ifdef block is (but now the ifdef is removed of course)
3. Remove a call to `PAL_Leave` in the `ASSERT` macro...this causes it to match what the `__GNUC__` block was doing.

These 3 extra changes tell me that this code path has probably been broken for a while and is not being tested/compiled.

File: src/pal/src/include/pal/dbgmsg.h [hunk 1]
```diff
@@ -322,32 +322,20 @@ bool DBG_ShouldCheckStackAlignment();
...
-   variable-argument macros */
-#ifdef __GNUC__
-
 /* define NOTRACE as nothing; this will absorb the variable-argument list used
    in tracing macros */
-#define NOTRACE(args...)
+#define NOTRACE(...)
 
 #if defined(__cplusplus) && defined(FEATURE_PAL_SXS)
 #define __ASSERT_ENTER()                                                \
```
Comment (marler8997) [inline src/pal/src/include/pal/dbgmsg.h]: Note: this macro was previously not in the `__STDC_VERSION__` ifdef block.

File: src/pal/src/include/pal/dbgmsg.h [hunk 2]
```diff
@@ -416,7 +361,6 @@ bool DBG_ShouldCheckStackAlignment();
     }                                                                   \
     if(g_Dbg_asserts_enabled)                                           \
     {                                                                   \
-        PAL_Leave();                                                    \
```
Comment (marler8997) [inline src/pal/src/include/pal/dbgmsg.h]: Note: removed this call to match what the `__GNUC__` block was doing.  Got an error when this is left in.

**Comment by marler8997** [comment]
On another note, the `DBG_printf_gcc` also included the function name whereas `DBG_printf_c99` does not.  This would mean losing some information by going to the current `c99` implementation.  Not sure if `__FUNCTION__` is not included because it's not supported or because it was a feature added to the `gcc` version that just wasn't added to the `c99` version? Should we add it to the `c99` version?

**Comment by janvorli** [comment]
@marler8997 I am sorry, I didn't know the two versions of the debug printing differed in more than the syntax for the variadic macro arguments and the flush count. What I would like to get here is what we had for the gcc version, but with the variadic macro arguments using the syntax that was in the c99 path.
I can see that the gcc version got some changes during the history that would be preferable to keep.

Could you please replace the body of the new DBG_printf_c99 by the body of the DBG_printf_gcc? Then we should be all set. I am sorry for the hassle again.

**Comment by marler8997** [comment]
Hey no problem, I don't mind making changes.  I'll go ahead and update the `c99` variant to match what the `gcc` variant is doing.

**Pull request merged**